File: lib/every/systemd.rb
module Every
  # Linux backend (beta): systemd user timers. Each task becomes a pair of
  # units in ~/.config/systemd/user — every-<name>.service (oneshot, invokes
  # `every run <name>`) and every-<name>.timer. Persistent=true mirrors
  # launchd's behavior of firing missed calendar runs on wake/boot.
  module Systemd
    UNIT_DIR = ENV["EVERY_SYSTEMD_DIR"] ? File.expand_path(ENV["EVERY_SYSTEMD_DIR"]) : Every.resolve_config_dir
    DAYS = %w[Sun Mon Tue Wed Thu Fri Sat].freeze

    module_function

    def unit_base(name)
      "every-#{name}"
    end

    def unit_path(name)
      File.join(UNIT_DIR, "#{unit_base(name)}.timer")
    end

    # Generic backend capability used by `doctor`.  File-backed schedulers can
    # answer this locally; registry/service-backed schedulers provide their own
    # implementation.
    def resource_exists?(name)
      File.exist?(unit_path(name))
    end

    def service_path(name)
      File.join(UNIT_DIR, "#{unit_base(name)}.service")
    end

    def write(name, _schedule = nil)
      schedule = _schedule
      FileUtils.mkdir_p(UNIT_DIR)
      File.write(service_path(name), service_unit(name))
      File.write(unit_path(name), timer_unit(name, schedule))
    end

    def service_unit(name)
      # systemd user services start from a clean environment; pin the resolved
      # data dir so the timer-spawned run reads the SAME store the task was
      # added under (EVERY_HOME or XDG_DATA_HOME won't be inherited).
      <<~UNIT
        [Unit]
        Description=every task #{name}

        [Service]
        Type=oneshot
        Environment=EVERY_HOME=#{DATA_DIR}
        ExecStart=#{RbConfig.ruby} "#{Runtime.bin}" run "#{name}"
      UNIT
    end

    def timer_unit(name, schedule)
      lines = ["[Unit]", "Description=every timer #{name}", "", "[Timer]"]
      # Tighten the default 1-minute slack so sub-minute intervals and calendar
      # times fire close to when launchd would on macOS.
      lines << "AccuracySec=1s"
      if schedule.kind == :interval
        lines << "OnActiveSec=#{schedule.interval}"
        lines << "OnUnitActiveSec=#{schedule.interval}"
      else
        calendar_lines(schedule).each { |c| lines << "OnCalendar=#{c}" }
        lines << "Persistent=true"
      end
      (lines + ["", "[Install]", "WantedBy=timers.target"]).join("\n") + "\n"
    end

    def calendar_lines(schedule)
      schedule.entries.map do |e|
        t = format("%02d:%02d:00", e["hour"], e["minute"])
        e["weekday"] ? "#{DAYS[e['weekday'] % 7]} *-*-* #{t}" : "*-*-* #{t}"
      end
    end

    def enable(name)
      sysctl("daemon-reload")
      out, st = sysctl("enable", "--now", "#{unit_base(name)}.timer")
      raise "systemctl enable failed: #{out.strip}" unless st.success?
      true
    end

    def disable(name)
      sysctl("disable", "--now", "#{unit_base(name)}.timer")
    end

    def loaded?(name)
      _out, st = sysctl("is-active", "--quiet", "#{unit_base(name)}.timer")
      st.success?
    end

    # All active every-timers in ONE call (vs one `is-active` per task).
    def loaded_names
      out, st = sysctl("list-units", "--type=timer", "--state=active",
                       "--no-legend", "--plain", "every-*.timer")
      return [] unless st.success?
      parse_units(out)
    end

    # `list-units --plain --no-legend` prints "UNIT LOAD ACTIVE SUB DESC" rows.
    def parse_units(out)
      out.lines.each_with_object([]) do |line, acc|
        # no-arg split discards leading blanks, so an indented row still parses
        m = line.split.first.to_s.match(/\Aevery-(.+)\.timer\z/)
        acc << m[1] if m
      end
    end

    def delete_units(name)
      [service_path(name), unit_path(name)].each do |p|
        File.delete(p) if File.exist?(p)
      end
      sysctl("daemon-reload")
    end

    def sysctl(*args)
      Open3.capture2e("systemctl", "--user", *args)
    end
  end
end

File: internal/backend/systemd.go
package backend

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/serhiileniv/every/internal/schedule"
)

// Systemd schedules through systemd user timers: a service and a timer per task.
type Systemd struct{ cfg Config }

func NewSystemd(cfg Config) *Systemd { return &Systemd{cfg: cfg} }

func (s *Systemd) Name() string { return "systemd" }

var systemdDays = [...]string{"Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"}

func (s *Systemd) unitBase(name string) string { return "every-" + name }

func (s *Systemd) UnitPath(name string) string {
	return filepath.Join(s.cfg.Dirs.Config, s.unitBase(name)+".timer")
}

func (s *Systemd) servicePath(name string) string {
	return filepath.Join(s.cfg.Dirs.Config, s.unitBase(name)+".service")
}

// ResourceExists checks the timer only: it is the unit that schedules, and a
// service without one is inert rather than half-registered.
func (s *Systemd) ResourceExists(name string) bool {
	_, err := os.Stat(s.UnitPath(name))
	return err == nil
}

func (s *Systemd) Write(name string, sched *schedule.Schedule) error {
	if err := os.MkdirAll(s.cfg.Dirs.Config, 0o755); err != nil {
		return err
	}
	if err := os.MkdirAll(s.cfg.Dirs.Logs, 0o755); err != nil {
		return err
	}
	if err := os.WriteFile(s.servicePath(name), []byte(s.ServiceUnit(name)), 0o644); err != nil {
		return err
	}
	return os.WriteFile(s.UnitPath(name), []byte(s.TimerUnit(name, sched)), 0o644)
}

// ServiceUnit is what the timer triggers.
//
// systemd user services start from a clean environment, so the resolved data
// dir is pinned: without it a timer-spawned run recomputes the default
// directory instead of reading the store the task was added under.
//
// The quoting is asymmetric -- the launcher and name are quoted, the
// Environment= value is not -- because that is what the Ruby emitted and the
// units are compared byte for byte.
func (s *Systemd) ServiceUnit(name string) string {
	return fmt.Sprintf(`[Unit]
Description=every task %s

[Service]
Type=oneshot
Environment=EVERY_HOME=%s
ExecStart=%s run "%s"
`, name, s.cfg.Dirs.Data, quoteUnlessPlaceholder(s.cfg.Launcher), name)
}

// quoteUnlessPlaceholder wraps the launcher in double quotes, which is how the
// path reaches systemd intact when it contains spaces.
func quoteUnlessPlaceholder(launcher string) string {
	return `"` + launcher + `"`
}

// TimerUnit schedules the service.
func (s *Systemd) TimerUnit(name string, sched *schedule.Schedule) string {
	lines := []string{"[Unit]", "Description=every timer " + name, "", "[Timer]"}

	// Tighten the default one-minute slack, so sub-minute intervals and
	// calendar times fire close to when launchd would on macOS.
	lines = append(lines, "AccuracySec=1s")

	if sched.Kind == schedule.Interval {
		lines = append(lines,
			fmt.Sprintf("OnActiveSec=%d", sched.Interval),
			fmt.Sprintf("OnUnitActiveSec=%d", sched.Interval))
	} else {
		for _, c := range CalendarLines(sched) {
			lines = append(lines, "OnCalendar="+c)
		}
		// Persistent mirrors launchd's behavior of firing missed calendar runs
		// on wake or boot.
		lines = append(lines, "Persistent=true")
	}

	lines = append(lines, "", "[Install]", "WantedBy=timers.target")
	return strings.Join(lines, "\n") + "\n"
}

// CalendarLines renders each entry as a systemd OnCalendar expression.
func CalendarLines(sched *schedule.Schedule) []string {
	out := make([]string, 0, len(sched.Entries))
	for _, e := range sched.Entries {
		t := fmt.Sprintf("%02d:%02d:00", e.Hour, e.Minute)
		if e.Weekday != nil {
			// The modulo clamps a legacy weekday 7 to Sunday rather than
			// indexing past the end of the table.
			out = append(out, fmt.Sprintf("%s *-*-* %s", systemdDays[((*e.Weekday%7)+7)%7], t))
			continue
		}
		out = append(out, "*-*-* "+t)
	}
	return out
}

func (s *Systemd) systemctl(args ...string) (string, error) {
	return runCmd("systemctl", append([]string{"--user"}, args...)...)
}

func (s *Systemd) Enable(name string) error {
	// The reload's result is deliberately ignored: on a host with no user bus
	// it fails, and the enable below reports that far more usefully.
	_, _ = s.systemctl("daemon-reload")

	out, err := s.systemctl("enable", "--now", s.unitBase(name)+".timer")
	if err != nil {
		return fmt.Errorf("systemctl enable failed: %s", strings.TrimSpace(out))
	}
	return nil
}

// Disable tolerates a timer the scheduler no longer has: rm disables before
// deleting, so failing here would strand the store entry.
func (s *Systemd) Disable(name string) error {
	_, _ = s.systemctl("disable", "--now", s.unitBase(name)+".timer")
	return nil
}

func (s *Systemd) DeleteUnits(name string) error {
	for _, p := range []string{s.servicePath(name), s.UnitPath(name)} {
		if err := os.Remove(p); err != nil && !os.IsNotExist(err) {
			return err
		}
	}
	_, _ = s.systemctl("daemon-reload")
	return nil
}

func (s *Systemd) Loaded(name string) bool {
	_, err := s.systemctl("is-active", "--quiet", s.unitBase(name)+".timer")
	return err == nil
}

func (s *Systemd) LoadedNames() ([]string, error) {
	out, err := s.systemctl("list-units", "--type=timer", "--state=active",
		"--no-legend", "--plain", "every-*.timer")
	if err != nil {
		// Soft failure, as on launchd: an unreachable systemd means nothing is
		// loaded, which doctor reports better than a crash would.
		return nil, nil
	}
	return parseUnits(out), nil
}

// parseUnits pulls task names out of `list-units --plain --no-legend` rows,
// which are "UNIT LOAD ACTIVE SUB DESCRIPTION".
var unitRe = regexp.MustCompile(`^every-(.+)\.timer$`)

func parseUnits(out string) []string {
	var names []string
	for _, line := range strings.Split(out, "\n") {
		// Whitespace-splitting discards leading blanks, so an indented row
		// still parses.
		fields := strings.Fields(line)
		if len(fields) == 0 {
			continue
		}
		if m := unitRe.FindStringSubmatch(fields[0]); m != nil {
			names = append(names, m[1])
		}
	}
	return names
}
